### DISCUSSION TITLE
Switch headers for each PID

### DISCUSSION THREAD

**Discussion by mateusleom** [VaAndCob/ESP32-Bluetooth-OBD2-Gauge#7] [Ideas]
I have a 2016 Subaru Forester XT with a CVT transmission. It is interesting to have some CVT information on the ESP32 OBDII Gauge, but the PIDs from the CVT come from the transmission control unit (TCU) while every other PID comes from the engine control unit (ECU).

It is possible to switch the address in which the ELM327 looks for the PIDs using the command AT SH = Set Header. The default header for every car is 7E0, which is the ECU. So if you just send any PID without specifying a header, 7E0 (ECU) will be the one answering. To get PIDs from the TCU, all you need to do is send the command AT SH 7E1 and then ELM327 will get pids from there, but it no longer talks to the ECU after that.

In order to get PIDs from ecu again, you need to send the command AT SH 7E0. 

So I have already managed to get PIDs from both the TCU and ECU, but not at the sime time, and thats what i need to do in order to have all PIDs i want in the ESP32 gauge. 

If anyone can help me modifying the code in order to do so, it will be greatly appreciated. 

Also, I've developed a 3d printed case that attachs to subarus AC vent, i'll make them available for everyone.

[Case Esp32.zip](https://github.com/user-attachments/files/17074395/Case.Esp32.zip)

**Comment by VaAndCob** [comment]
> I have a 2016 Subaru Forester XT with a CVT transmission. It is interesting to have some CVT information on the ESP32 OBDII Gauge, but the PIDs from the CVT come from the transmission control unit (TCU) while every other PID comes from the engine control unit (ECU).
> 
> It is possible to switch the address in which the ELM327 looks for the PIDs using the command AT SH = Set Header. The default header for every car is 7E0, which is the ECU. So if you just send any PID without specifying a header, 7E0 (ECU) will be the one answering. To get PIDs from the TCU, all you need to do is send the command AT SH 7E1 and then ELM327 will get pids from there, but it no longer talks to the ECU after that.
> 
> In order to get PIDs from ecu again, you need to send the command AT SH 7E0.
> 
> So I have already managed to get PIDs from both the TCU and ECU, but not at the sime time, and thats what i need to do in order to have all PIDs i want in the ESP32 gauge.
> 
> If anyone can help me modifying the code in order to do so, it will be greatly appreciated.
> 
> Also, I've developed a 3d printed case that attachs to subarus AC vent, i'll make them available for everyone.
> 
> [Case Esp32.zip](https://github.com/user-attachments/files/17074395/Case.Esp32.zip)

thank you for your case design.

**Reply by mateusleom** [reply]
Is it possible to code something as

If (PID == "2210D2")
     {
     sendToElm("ATSH7E0");
     }
else
     {
     sendToElm("ATSH7E1");
     }

?

This need to be sent before asking for the pid, so the code will be able to get info from 7E1 and 7E0 headers at the same time.
I've been trying to do something like this but haven't had any lucky so far. Can you kindly help?

**Reply by VaAndCob** [reply]
> Is it possible to code something as
> 
> If (PID == "2210D2") { sendToElm("ATSH7E1"); } else { sendToElm("ATSH7E1"); }
> 
> ?
> 
> This need to be sent before asking for the pid, so the code will be able to get info from 7E1 and 7E0 headers at the same time. I've been trying to do something like this but haven't had any lucky so far. Can you kindly help?

if I have time, I will dig into how to work ELM327. oh you can read the manual here [ELM327](https://www.elmelectronics.com/DSheets/ELM327DSH.pdf)

**Reply by VaAndCob** [reply]
> Is it possible to code something as
> 
> If (PID == "2210D2") { sendToElm("ATSH7E1"); } else { sendToElm("ATSH7E1"); }
> 
> ?
> 
> This need to be sent before asking for the pid, so the code will be able to get info from 7E1 and 7E0 headers at the same time. I've been trying to do something like this but haven't had any lucky so far. Can you kindly help?

yes, you need to switch back and forth. like that. if I have free time, I will try do it.

**Reply by mateusleom** [reply]
> > Is it possible to code something as
> > If (PID == "2210D2") { sendToElm("ATSH7E1"); } else { sendToElm("ATSH7E1"); }
> > ?
> > This need to be sent before asking for the pid, so the code will be able to get info from 7E1 and 7E0 headers at the same time. I've been trying to do something like this but haven't had any lucky so far. Can you kindly help?
> 
> yes, you need to switch back and forth. like that. if I have free time, I will try do it.

I read it all and managed to get it working, but not both headers at the same time.  If i modify line #68 and #69 to

const uint8_t elm327InitCount = 9;
const String elm327Init[elm327InitCount] = { "ATZ", "ATBRD23", "ATSP0", "ATAT2", "ATL0", "ATH0", "ATE0", "0100", "ATSH7E1" };  //agressive wait response time from ECU

It will read pids from 7E1, but will cease communication to 7E0.
